include/circt/Dialect/SMT/SMTBitVectorOps.td

```diff
@@ -61,4 +61,50 @@ def BVConstantOp : SMTBVOp<"constant", [
   let hasFolder = true;
 }
 
+class BVArithmeticOrBitwiseOp<string mnemonic, string desc> :
+    SMTBVOp<mnemonic, [Pure, SameOperandsAndResultType]> {
```

Comment:
Can you add the `ConditionallySpeculatable` trait? That's just a trait to say that this operation has no undefined behavior, so you can hoist it out of a loop safely (which Pure doesn't let you do).

Reply:
`Pure` means speculatable and memory-effect-free. See the definitions from the tablegen file:
```
def AlwaysSpeculatable : TraitList<[ConditionallySpeculatable, AlwaysSpeculatableImplTrait]>;
def Pure : TraitList<[AlwaysSpeculatable, NoMemoryEffect]>;
```

Reply:
Oh, my bad, I thought it was distinct!